postcss/theme.js

```diff
@@ -0,0 +1,180 @@
+const { Declaration, AtRule, Rule } = require('postcss');
+const supportedThemes = ['dark', 'light'];
+
+const getPropertyName = (selector, decl) => {
+    const regex = /([^A-Za-z0-9\-_])/g
+    return `--${selector}_${decl.prop}`
+        .replace(/\s+/g, '_')
+        .replace(regex, '\\$1')
+}
+
+const splitRule = (rule, selectorToExtract) => {
+    // |cloneAfter|, so in an |each| loop, the new rule will be processed.
+    const cloned = rule.cloneAfter();
+
+    // Remove the light selector from the cloned rule.
+    // Note: Directly splicing |selectors| has no effect, we have
+    // to set a varible.
+    const index = cloned.selectors.indexOf(selectorToExtract);
+    const copy = [...cloned.selectors];
+    copy.splice(index, 1);
+    cloned.selectors = copy;
+
+    // If we don't set this all the formatting breaks.
+    cloned.raws.before = '\n';
+
+    // Set the selectors on the light rule to be just the light selector.
+    rule.selector = selectorToExtract;
+}
+
+const defaultOptions = {
+    darkSelector: '[data-theme=dark]',
+    lightSelector: '[data-theme=light]'
+}
+
+/**
+ * @param {{ darkSelector: string, lightSelector: string }} options The options for configuring the selectors for darkmode.
+ */
+module.exports = (options) => {
+    options = { ...defaultOptions, ...options };
+
+    const rules = {};
+    const nodesToDelete = new Set();
+    const findMatchingBaseRules = root => {
+        const selectors = new Set(Object.keys(rules));
+        root.each(rule => {
+            if (rule.type !== 'rule') return;
+
+            for (const baseSelector of [...rule.selectors]) {
+                if (!selectors.has(baseSelector)) continue
+
+                // If the rule has multiple selectors, we need to split the the
+                // rule apart:
+                // For example, when processing the rule .frob:
+                // .foo, .frob, .bar { background: red; }
+                // should become
+                // .frob { background: red; }
+                // .foo, .bar { background: red; }
```

Comment:
Why would it become 2 rules instead of three, given that there are three selectors?

Reply:
Because at this point, we only know that `.frob` is themed (it has a light/dark variation), so we might as well leave `.foo, .bar` as a combined selector rather than duplicating it and making a larger CSS file. If we encounter a theme variation for either of them later we can split them then.

This function splits the themed selector out from the other ones.